paddle/optimizer/parameter_optimizer.h

```diff
@@ -0,0 +1,54 @@
+#ifndef PADDLE_PARAMETER_OPTIMIZER_H_
+#define PADDLE_PARAMETER_OPTIMIZER_H_
+
+#include <glog/logging.h>
+#include <functional>
+#include <string>
+#include "Tensor.h"
+#include "OptimizerConfig.pb.h"
+
+namespace paddle {
+namespace optimizer {
+
+template <class T>
+class ParameterOptimizer {
+public:
+  /**
+   * @brief  update hook for algorithm need to traverse parameter more than
+   * once.
+   */
+  ParameterOptimizer(const OptimizerConfig &config) : config_(config){};
+
+  static ParameterOptimizer *create(const ::std::string &config_proto);
+  virtual void update(const Tensor &gradient) = 0;
+  virtual void destroy() = 0;
+  virtual T *get_weight() const;
+  virtual void set_weight(const Tensor<T> *parameter);
+  // package optimizer config proto in runtime for saving checkpoint
+  virtual char* get_config_proto();
```

Comment:
same with last comment, I'm agree with you that optimizer should better leaving the parsing job to the create interface. But we need something to pack training state back.